### DISCUSSION TITLE
Proxy with authentication?

### DISCUSSION THREAD

**Discussion by lukebelkis** [ultrafunkamsterdam/undetected-chromedriver#1226] [Q&A] [answered]
I have a proxy with authentication and I'd like to use it on uc. 

how can I do this? Thanks!

**Comment by bberkay** [comment, marked as answer]
I had the same problem and I solved it like this with python:
```
from selenium.webdriver.common.proxy import Proxy, ProxyType
from selenium import webdriver
import undetected_chromedriver as uc 

proxy = Proxy()
proxy.proxy_type = ProxyType.MANUAL
proxy.http_proxy = "username:password@ip:port"
proxy.ssl_proxy = "username:password@ip:prt"
capabilities = webdriver.DesiredCapabilities.CHROME.copy()
proxy.add_to_capabilities(capabilities)

driver = uc.Chrome(use_subprocess=True, desired_capabilities=capabilities)
```

**Reply by lukebelkis** [reply]
Thank you very much!

**Reply by silverxd20** [reply]
anyone else has tested this answer? it does not work for me :/

**Reply by bberkay** [reply]
> anyone else has tested this answer? it does not work for me :/

As another method, you can use seleniumwire and undetected_chromedriver at the same time.
```
import seleniumwire.undetected_chromedriver as uc

proxy_options = {
    'proxy': {
        'http': "http://username:password@host:port",
        'https': "http://username:password@host:port"
    }
}

driver = uc.Chrome(use_subprocess=True, seleniumwire_options=proxy_options)
```

**Reply by silverxd20** [reply]
this code works but the page load without https, this maybe because we are using "selenium-wire"? bec![Screenshot_1](https://github.com/ultrafunkamsterdam/undetected-chromedriver/assets/13056146/a35136f9-8eba-47fa-ad32-746b2e2bdbb5)

**Reply by Noodles0x00** [reply]
seleniumwire acts as a proxy itself, this should only be visual for the most part, it's the same as using burp suite for example

**Reply by BRKind** [reply]
thank you so much

**Reply by Naikasinji93** [reply]
Hi @bberkay I reused:

proxy = Proxy()
proxy.proxy_type = ProxyType.MANUAL
proxy.http_proxy = "username:password@ip:port"
proxy.ssl_proxy = "username:password@ip:prt"
capabilities = webdriver.DesiredCapabilities.CHROME.copy()

But I don't have an `add_to_capabilities()` function, the system recommended `to_capabilities()` instead, which does not take a single argument. Could you help me out?

**Reply by lynnwho12** [reply]
same issue

**Comment by nullANzero** [comment]
class ChromeProxyGenerator:
    def __init__(self, proxy_host, proxy_port, proxy_user, proxy_pass, proxy_folder):
        self.proxy_host = proxy_host
        self.proxy_port = proxy_port
        self.proxy_user = proxy_user
        self.proxy_pass = proxy_pass
        self.proxy_folder = proxy_folder

        """
        proxy_host = "example.com"
        proxy_port = 8080
        proxy_user = "username"
        proxy_pass = "password"
        proxy_folder = "/path/to/proxy/folder"
        
        generator = ChromeProxyGenerator(proxy_host, proxy_port, proxy_user, proxy_pass, proxy_folder)
        generator.proxy_connection()
        """

    def proxy_connection(self):
        manifest_json = """
            {
                "version": "1.0.0",
                "manifest_version": 2,
                "name": "Chrome Proxy",
                "permissions": [
                    "proxy",
                    "tabs",
                    "unlimitedStorage",
                    "storage",
                    "<all_urls>",
                    "webRequest",
                    "webRequestBlocking"
                ],
                "background": {
                    "scripts": ["background.js"]
                },
                "minimum_chrome_version":"22.0.0"
            }
        """

        background_js = """
            var config = {
                    mode: "fixed_servers",
                    rules: {
                    singleProxy: {
                        scheme: "http",
                        host: "%s",
                        port: parseInt(%s)
                    },
                    bypassList: ["localhost"]
                    }
                };
            chrome.proxy.settings.set({value: config, scope: "regular"}, function() {});
            function callbackFn(details) {
                return {
                    authCredentials: {
                        username: "%s",
                        password: "%s"
                    }
                };
            }
            chrome.webRequest.onAuthRequired.addListener(
                        callbackFn,
                        {urls: ["<all_urls>"]},
                        ['blocking']
            );
            """ % (self.proxy_host, self.proxy_port, self.proxy_user, self.proxy_pass)

        with open(f"{self.proxy_folder}/manifest.json", "w") as f:
            f.write(manifest_json)
        with open(f"{self.proxy_folder}/background.js", "w") as f:
            f.write(background_js)

> > кто-нибудь еще тестировал этот ответ? у меня он не работает :/
> 
> В качестве другого метода вы можете использовать seleniumwire и undetected_chromedriver одновременно.
> 
> ```
> import seleniumwire.undetected_chromedriver as uc
> 
> proxy_options = {
>     'proxy': {
>         'http': "http://username:password@host:port",
>         'https': "http://username:password@host:port"
>     }
> }
> 
> driver = uc.Chrome(use_subprocess=True, seleniumwire_options=proxy_options)
> ```

**Reply by dxdate** [reply]
how to use socks5 proxy with authentication?
